### DISCUSSION TITLE
Bring 3D tiles from cesium into the ITwin app

### DISCUSSION THREAD

**Discussion by salah-walid** [iTwin/itwinjs-core#3853] [Q&A]
Hi,

We have been trying to bring cesium Tiles to the ITwin app using this code 
```typescript
const cesiumProps = {
  tilesetUrl: getCesiumAssetUrl(
    cesiumId, 
    process.env.IMJS_CESIUM_KEY ?? ""
  )
};
viewPort.displayStyle.attachRealityModel(cesiumProps);
```

It used to work with ITwin v2 but not anymore in ITwin V3, Do we have to use context campture instead ?

**Comment by aruniverse** [comment]
Are you initializing your TileAdmin with a cesium ion key? 
https://www.itwinjs.org/reference/core-frontend/tiles/tileadmin/tileadmin.props/cesiumionkey/
https://www.itwinjs.org/changehistory/3.0.0/#removed-default-api-keys

**Reply by salah-walid** [reply]
Yes we added it to the Viewer component prop directly like so:
```typescript
<Viewer
    iTwinId={iTwinId}
    iModelId={iModelId}
    authClient={myTokenServerAuthClient}
    enablePerformanceMonitors={false}
    viewCreatorOptions={viewCreatorOptions}
    mapLayerOptions={mapLayerOptions }
    tileAdminOptions={{ cesiumIonKey: cesiumKey }}
>
```
Still the tile is not showing

**Reply by pmconne** [reply]
Did you try replacing `process.env.IMJS_CESIUM_KEY` with `IModelApp.tileAdmin.cesiumIonKey` (or simply passing your key directly)?
Did you look at the network activity in browser dev tools to see what is happening with the request?

**Reply by salah-walid** [reply]
Yes for all the above, also a weird thing that i noted is that the url returned by the `getCesiumAssetUrl` is in the format of
`"$CesiumIonAsset={asset_id}:{cesiumIonKey}"` which doesn't look like a typical url, is that normal ?

**Reply by pmconne** [reply]
I don't know. Looks like [kind of a hack](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/tile/CesiumAssetProvider.ts)? @MarcBedard8?

**Reply by MarcBedard8** [reply]
I agree that the name of the method is misleading. It actually returns a string that can later be converted into an URL.
It will later be converted into:
  const requestTemplate = `https://api.cesium.com/v1/assets/${assetId}/endpoint?access_token={CesiumRequestToken}`;
  const apiUrl: string = requestTemplate.replace("{CesiumRequestToken}", requestKey);

You should see this call to https://api.cesium.com/v1/assets... in your network tab in your debug window.
Response  to this call might provide  more clue about  what is happening here.

**Reply by salah-walid** [reply]
I have this response in the debug dialog
```
{
    "type": "3DTILES",
    "url": "https://assets.cesium.com/{id}/tileset.json?v=4",
    "accessToken": "{token}",
    "attributions": [
        {
            "html": "<span><a href=\"https://cesium.com\" target=\"_blank\"><img alt=\"Cesium ion\" src=\"https://assets.cesium.com/ion-credit.png\"></a></span>",
            "collapsible": false
        },
        {
            "html": "<a href=\"https://cesium.com/pricing/\" target=\"_blank\">Upgrade for commercial use.</a>",
            "collapsible": false
        }
    ]
}
```

**Reply by MarcBedard8** [reply]
@pmconne , This seems like we reach the site, but I haven't use much cesium asset. Is "Upgrade for commercial use" means a new keys is required?

**Reply by pmconne** [reply]
Seem more likely that [monthly usage limits](https://cesium.com/platform/cesium-ion/pricing/) have been exceeded.
@salah-walid I think you need to consult Cesium ION documentation and/or support.

**Reply by salah-walid** [reply]
Using the cesium sdk the tiles work perfectly, i don't think it's a problem with cesium

**Reply by salah-walid** [reply]
Do i need to use that url returned instead?

**Reply by pmconne** [reply]
No, we should be fetching that URL ourselves after fetching api.cesium.com.
Are you seeing a subsequent request to that URL? If so, what is the response?

The attributions in the response to my own  request to api.cesium.com do not mention upgrading for commercial use, but maybe they just always include that for non-commercial users.

```
{
  "accessToken": "{my-access-token}",
  "attributions": [
    {
      "collapsible": false,
      "html": "<span><a href=\"https://cesium.com\" target=\"_blank\"><img alt=\"Cesium ion\" src=\"https://assets.cesium.com/ion-credit.png\"></a></span>"
    },
    {
      "collapsible": true,
      "html": "<span>© <a href=\"https://www.openstreetmap.org/copyright\" target=\"_blank\">OpenStreetMap</a> contributors</span>"
    }
  ],
  "type": "3DTILES",
  "url": "https://assets.cesium.com/{my-asset-id}/tileset.json?v=15"
}
```

**Reply by salah-walid** [reply]
I don't see any subsequent requests in the network tab to `https://assets.cesium.com/{my-asset-id}/tileset.json?v=4`

**Reply by pmconne** [reply]
@MarcBedard8 theories or suggestions?

**Reply by MarcBedard8** [reply]
Could it be that the location of the asset you are trying to display is not at the location of your iModel? Or maybe your iModel is not geolocated?

**Reply by MarcBedard8** [reply]
I might have found the problem. 
It seems we are forcing a predefined assetId (OSM Building) here:
https://github.com/iTwin/itwinjs-core/blob/a49739fd86c80f58a2e12a682fb53ed0d7895613/core/frontend/src/RealityDataSource.ts#L120

I'll work on a fix and will let you know.

**Reply by MarcBedard8** [reply]
See this PR: https://github.com/iTwin/itwinjs-core/pull/3885

**Reply by salah-walid** [reply]
After updating my dependencies and testing, still nothing showing on the viewer
Here are my dependencies
```TypeScript
"dependencies": {
    "@bentley/icons-generic": "^1.0.34",
    "@bentley/react-scripts": "^4.0.7",
    "@itwin/appui-abstract": "3.2.9",
    "@itwin/appui-layout-react": "3.2.9",
    "@itwin/appui-react": "3.2.9",
    "@itwin/browser-authorization": "^0.5.1",
    "@itwin/components-react": "3.2.9",
    "@itwin/core-bentley": "3.2.9",
    "@itwin/core-common": "3.2.9",
    "@itwin/core-frontend": "3.2.9",
    "@itwin/core-geometry": "3.2.9",
    "@itwin/core-i18n": "3.2.9",
    "@itwin/core-markup": "3.2.9",
    "@itwin/core-orbitgt": "3.2.9",
    "@itwin/core-quantity": "3.2.9",
    "@itwin/core-react": "3.2.9",
    "@itwin/core-telemetry": "3.2.9",
    "@itwin/imodel-components-react": "3.2.9",
    "@itwin/imodels-access-frontend": "^1.0.1",
    "@itwin/imodels-client-management": "^1.0.1",
    "@itwin/itwinui-react": "^1.36.0",
    "@itwin/presentation-common": "3.2.9",
    "@itwin/presentation-frontend": "3.2.9",
    "@itwin/reality-data-client": "^0.7.2",
    "@itwin/web-viewer-react": "^2.0.9",
    "@itwin/webgl-compatibility": "3.2.9",
    "@types/history": "^4.7.11",
    "@types/node": "^14.18.13",
    "@types/react": "^17.0.44",
    "@types/react-dom": "^17.0.15",
    "history": "^4.10.1",
    "mobx": "^6.6.1",
    "mobx-react": "^7.5.2",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "typescript": "~4.3.0"
  },
```
And here is the results returned by the vp.displayStyle.attachRealityModel(cesiumProps) method

![Screenshot_2](https://user-images.githubusercontent.com/29980658/197874433-3c8c87b4-913a-4927-9ce9-ccf8d3c12667.png)
